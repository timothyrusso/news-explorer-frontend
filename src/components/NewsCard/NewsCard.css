.card {
  width: 100%;
  border-radius: 15px;
  overflow: hidden;
  position: relative;
}

.card__favourite {
  max-width: 40px;
  min-height: 40px;
  width: 100%;
  position: absolute;
  border: none;
  background-color: #fff;
  top: 18px;
  right: 15px;
  border-radius: 8px;
  background-image: url(../../images/bookmark.svg);
  background-repeat: no-repeat;
  background-position: center center;
  cursor: pointer;
}

.card__image {
  min-height: 272px;
  background-size: cover;
  cursor: pointer;
}

.card__content {
  background-color: #fff;
  box-sizing: border-box;
  padding: 20px 24px 24px;
  min-height: 304px;
  height: 100%;
}

.card__date {
  font-family: 'Inter', Arial, sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 18px;
  line-height: 1.33;
  color: #B6BCBF;
  margin: 0;
}

.card__title {
  font-family: 'Roboto Slab';
  font-style: normal;
  font-weight: 400;
  font-size: 26px;
  line-height: 1.15;
  color: #1A1B22;
  margin: 12px 0 16px;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

.card__text {
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.38;
  color: #1A1B22;
  margin: 0;
  display: -webkit-box;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

.card__source {
  font-family: 'Roboto Slab';
  font-style: normal;
  font-weight: 700;
  font-size: 16px;
  line-height: 1.25;
  letter-spacing: 0.4px;
  text-transform: uppercase;
  color: #B6BCBF;
  margin: 0;
  display: -webkit-box;
  -webkit-line-clamp: 1;
  -webkit-box-orient: vertical;
  overflow: hidden;
  position: absolute;
  bottom: 24px;
  left: 24px;
}

.show-more-button {
  border: none;
  background-color: #fff;
  max-width: 288px;
  width: 100%;
  min-height: 64px;
  border-radius: 80px;
  display: block;
  margin: 0 auto;
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 500;
  font-size: 18px;
  line-height: 1.33;
  color: #1A1B22;
}
