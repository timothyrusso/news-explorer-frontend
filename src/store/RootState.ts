import { UserState } from './user/user.reducer';
import { ArticleState } from './article/article.reducer';
import { TogglesState } from './toggles/toggles.reducer';
import { ErrorsState } from './errors/errors.reducer';

export type RootState = {
  user: UserState;
  article: ArticleState;
  toggles: TogglesState;
  errors: ErrorsState;
};
