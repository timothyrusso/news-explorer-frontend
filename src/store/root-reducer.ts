import { combineReducers } from 'redux';
import { userReducer } from './user/user.reducer';
import { articleReducer } from './article/article.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  article: articleReducer,
});
