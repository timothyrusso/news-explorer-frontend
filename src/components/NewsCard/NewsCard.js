import "./NewsCard.css";
import alternativeBackground from "../../images/header_background.png";
import { useState, useEffect } from "react";

const NewsCard = ({
  card,
  onSigninPopupClick,
  loggedIn,
  savedCard,
  handleBookmarkClick,
  handleDeleteArticles,
  checkSavedArticle,
  saveUnauthorizedUserCard,
}) => {
  const [bookmarkStatus, setBookmarkStatus] = useState(false);

  const saveCard = () => {
    handleBookmarkClick(card);
  };

  const deleteCard = () => {
    handleDeleteArticles(card);
  };

  const handleCardSaveUnauthorizedUser = () => {
    saveUnauthorizedUserCard(card);
    onSigninPopupClick();
  };

  const convertDate = (timeStr) => {
    var date = new Date(timeStr);
    var day = date.getDate();
    var year = date.getFullYear();
    var month = date.getMonth() + 1;

    date.setMonth(month - 1);
    const monthName = date.toLocaleString("default", { month: "long" });

    var dateStr = monthName + " " + day + ", " + year;

    return dateStr;
  };

  const data = convertDate(!savedCard ? card.publishedAt : card.date);

  const backgroundUrl = !savedCard ? card.urlToImage : card.image;

  useEffect(() => {
    if (checkSavedArticle(card)) {
      setBookmarkStatus(true);
    } else {
      setBookmarkStatus(false);
    }
  }, [loggedIn, card, checkSavedArticle]);

  return (
    <li className="card">
      {savedCard && (
        <>
          <div className="card__keyword">{card.keyword}</div>
          <button
            aria-label="Delete"
            type="button"
            className="card__delete"
            onClick={deleteCard}
          ></button>
          <dialog className="card__tooltip card__tooltip_type_delete">
            Remove from saved
          </dialog>
        </>
      )}
      {!savedCard && (
        <>
          <button
            aria-label="Favourite"
            type="button"
            className={`card__favourite ${
              bookmarkStatus ? "card__favourite_active" : ""
            }`}
            onClick={!loggedIn ? handleCardSaveUnauthorizedUser : saveCard}
          ></button>
          {!loggedIn && (
            <dialog className="card__tooltip">Sign in to save articles</dialog>
          )}
        </>
      )}
      <a
        href={!savedCard ? card.url : card.link}
        target="_blank"
        rel="noreferrer"
      >
        <div
          className="card__image"
          style={{
            backgroundImage: `url(${
              backgroundUrl != null ? backgroundUrl : alternativeBackground
            })`,
          }}
          href={!savedCard ? card.url : card.link}
        ></div>
      </a>
      <div className="card__content">
        <time className="card__date">{data}</time>
        <a
          href={!savedCard ? card.url : card.link}
          target="_blank"
          rel="noreferrer"
          className="card__link-wrapper"
        >
          <h3 className="card__title">{card.title}</h3>
        </a>
        <a
          href={!savedCard ? card.url : card.link}
          target="_blank"
          rel="noreferrer"
          className="card__link-wrapper"
        >
          <p className="card__text">{!savedCard ? card.content : card.text}</p>
        </a>
        <a
          href={!savedCard ? card.url : card.link}
          target="_blank"
          rel="noreferrer"
          className="card__link-wrapper"
        >
          <p className="card__source">
            {!savedCard ? card.source.name : card.source}
          </p>
        </a>
      </div>
    </li>
  );
};

export default NewsCard;
