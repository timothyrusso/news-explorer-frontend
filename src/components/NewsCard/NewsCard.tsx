import React, { FC } from 'react';
import alternativeBackground from '../../images/header_background.png';
import { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { usePopup } from '../../hooks/usePopup';
import { useHandleBookmarkClick } from '../../hooks/useHandleBookmarkClick';
import { useArticleApi } from '../../hooks/useArticleApi';
import { setTemporarySavedArticleAction } from '../../store/article/article.actions';
import { useSelector } from 'react-redux';
import { RootState } from '../../store/RootState';
import { Article } from '../../store/article/article.type';
import { SavedArticle } from '../../store/article/article.type';
import './NewsCard.css';

type NewsCardProps = {
  card: SavedArticle | Article;
};

const NewsCard: FC<NewsCardProps> = ({ card }) => {
  const [bookmarkStatus, setBookmarkStatus] = useState(false);

  const { handleSigninPopupClick } = usePopup();
  const { handleBookmarkClick } = useHandleBookmarkClick();
  const { handleDeleteArticles, checkSavedArticle } = useArticleApi();

  const isSavedArticle = useSelector(
    (state: RootState) => state.article.isSavedArticle
  );
  const loggedIn = useSelector((state: RootState) => state.toggles.isLoggedin);

  const dispatch = useDispatch();

  const saveCard = () => {
    if ('url' in card) handleBookmarkClick(card);
  };

  const deleteCard = () => {
    if ('_id' in card) handleDeleteArticles(card);
  };

  const handleCardSaveUnauthorizedUser = () => {
    if ('url' in card) dispatch(setTemporarySavedArticleAction([card]));
    handleSigninPopupClick();
  };

  const convertDate = (timeStr: string) => {
    var date = new Date(timeStr);
    var day = date.getDate();
    var year = date.getFullYear();
    var month = date.getMonth() + 1;

    date.setMonth(month - 1);
    const monthName = date.toLocaleString('default', { month: 'long' });

    var dateStr = monthName + ' ' + day + ', ' + year;

    return dateStr;
  };

  const data = convertDate(
    !isSavedArticle && 'publishedAt' in card
      ? card.publishedAt
      : (card as SavedArticle).date
  );

  const backgroundUrl =
    !isSavedArticle && 'urlToImage' in card
      ? card.urlToImage
      : (card as SavedArticle).image;

  useEffect(() => {
    if ('url' in card && checkSavedArticle(card)) {
      setBookmarkStatus(true);
    } else {
      setBookmarkStatus(false);
    }
  }, [loggedIn, card, checkSavedArticle]);

  return (
    <li className="card">
      {isSavedArticle && (
        <>
          <div className="card__keyword">
            {'keyword' in card && card.keyword}
          </div>
          <button
            aria-label="Delete"
            type="button"
            className="card__delete"
            onClick={deleteCard}
          ></button>
          <dialog className="card__tooltip card__tooltip_type_delete">
            Remove from saved
          </dialog>
        </>
      )}
      {!isSavedArticle && (
        <>
          <button
            aria-label="Favourite"
            type="button"
            className={`card__favourite ${
              bookmarkStatus ? 'card__favourite_active' : ''
            }`}
            onClick={!loggedIn ? handleCardSaveUnauthorizedUser : saveCard}
          ></button>
          {!loggedIn && (
            <dialog className="card__tooltip">Sign in to save articles</dialog>
          )}
        </>
      )}
      <a
        href={
          !isSavedArticle && 'url' in card
            ? card.url
            : (card as SavedArticle).link
        }
        target="_blank"
        rel="noreferrer"
      >
        <div
          className="card__image"
          style={{
            backgroundImage: `url(${
              backgroundUrl != null ? backgroundUrl : alternativeBackground
            })`,
          }}
          // @ts-ignore
          href={
            !isSavedArticle && 'url' in card
              ? card.url
              : (card as SavedArticle).link
          }
        ></div>
      </a>
      <div className="card__content">
        <time className="card__date">{data}</time>
        <a
          href={
            !isSavedArticle && 'url' in card
              ? card.url
              : (card as SavedArticle).link
          }
          target="_blank"
          rel="noreferrer"
          className="card__link-wrapper"
        >
          <h3 className="card__title">{card.title}</h3>
        </a>
        <a
          href={
            !isSavedArticle && 'url' in card
              ? card.url
              : (card as SavedArticle).link
          }
          target="_blank"
          rel="noreferrer"
          className="card__link-wrapper"
        >
          <p className="card__text">
            {!isSavedArticle && 'content' in card
              ? card.content
              : (card as SavedArticle).text}
          </p>
        </a>
        <a
          href={
            !isSavedArticle && 'url' in card
              ? card.url
              : (card as SavedArticle).link
          }
          target="_blank"
          rel="noreferrer"
          className="card__link-wrapper"
        >
          <p className="card__source">
            {!isSavedArticle && typeof card.source !== 'string'
              ? card.source.name
              : card.source}
          </p>
        </a>
      </div>
    </li>
  );
};

export default NewsCard;
