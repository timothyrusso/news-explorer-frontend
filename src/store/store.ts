import { legacy_createStore as createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './root-reducer';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { RootState } from './RootState';

const store = createStore(rootReducer, applyMiddleware(thunk));

export type AppDispatch = ThunkDispatch<RootState, unknown, AnyAction>;

export default store;
