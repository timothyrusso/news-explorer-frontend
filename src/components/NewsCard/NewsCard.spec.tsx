import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../store/store';
import NewsCard from './NewsCard';

const SavedArticle = {
  _id: '_id',
  keyword: 'string',
  title: 'string',
  text: 'string',
  date: 'dd-mm-yyyy',
  source: 'string',
  link: 'string',
  image: 'string',
  __v: 1,
};

describe('NewsCard component', () => {
  it('Should render the NewsCard component', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <NewsCard key="_id" card={SavedArticle} />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
