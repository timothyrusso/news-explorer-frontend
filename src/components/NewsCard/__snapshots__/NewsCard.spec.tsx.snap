// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsCard component Should render the NewsCard component 1`] = `
<DocumentFragment>
  <li
    class="card"
  >
    <button
      aria-label="Favourite"
      class="card__favourite "
      type="button"
    />
    <dialog
      class="card__tooltip"
    >
      Sign in to save articles
    </dialog>
    <a
      href="string"
      rel="noreferrer"
      target="_blank"
    >
      <div
        class="card__image"
        href="string"
        style="background-image: url(string);"
      />
    </a>
    <div
      class="card__content"
    >
      <time
        class="card__date"
      >
        Invalid Date NaN, NaN
      </time>
      <a
        class="card__link-wrapper"
        href="string"
        rel="noreferrer"
        target="_blank"
      >
        <h3
          class="card__title"
        >
          string
        </h3>
      </a>
      <a
        class="card__link-wrapper"
        href="string"
        rel="noreferrer"
        target="_blank"
      >
        <p
          class="card__text"
        >
          string
        </p>
      </a>
      <a
        class="card__link-wrapper"
        href="string"
        rel="noreferrer"
        target="_blank"
      >
        <p
          class="card__source"
        >
          string
        </p>
      </a>
    </div>
  </li>
</DocumentFragment>
`;
